Use Melissa E. O'Neill's pcg for rng and Lemire's unbiased bounded rand

Andersama opened the pull request:
Mersenne Twister's state size is huge (kb's of state), instead we'll swap with a pcg. The pcg's state is 128 bits and has a equidistributed 2^64 period (plenty for a quick rng). To replicate std::distribution we use lemire's unbiased bounded random method.

Additionally we seed using the address of the rng struct, in theory this is randomly allocated somewhere, but since this is a tiny device with not a lot of of address space, this likely only flips the lower bits. It still uses the motion accelerometers 3-axis data as before, but uses the three uint32_t results to create one uint64_t. The rng shuffles itself a bit further by replacing its state using the initial seeding.

mark9064 commented:
https://www.youtube.com/watch?v=ia8Q51ouA_s&t=78

Seriously though, this looks great :)  really good spot that the mersenne twister is way too big for this application (and that's probably the cause of a crash I've been seeing recently)

I'm not too experienced with random number generation techniques so I need to read over them properly to review, but at a quick glance this looks very promising

Andersama commented:
@Avamander that's the bluetooth device right? I did a quick search and spotted that there's an RNG interrupt, so I assume there's some form of secure rng available. I would use it, if I knew how.

Andersama commented:
@mark9064 I believe looking at `ble_ll_rand` that realistically we're grabbing random data using `ble_ll_rand_data_get((uint8_t *)xsubi, sizeof(xsubi));`? It seems like `jrand48` is a transform to take the random data from 48 bits of random data to output 32. We could just use `ble_ll_rand_data_get` to fill the pcg fairly easily.

Andersama commented:
@Avamander the cost of the pcg should only be 128 bits of memory to hold the state at most (depends on the size choice of the implementation) for any instance being used (or double that in my rough conceptual model where there were two at work, one for the screen, one to generate new seeds) and only a few bytes for the instructions for the two functions.

Andersama committed:
Use Melissa E. O'Neill's pcg for rng and Lemire's unbiased bounded rand

    Mersenne Twister's state size is huge (kb's of state), instead we'll swap with a pcg.
    The pcg's state is n bits and has a equidistributed 2^(n/2) period (plenty for a quick rng).
    To replicate std::distribution we use lemire's unbiased bounded random method.

    Adjust the state size as desired between 32 and 128 bits.

Andersama commented:
Thoroughly confused myself with undefined reference problems. I'm leaving as is, though feel free to change the state size, it should adjust as desired, though I'd recommend sticking with the 128 bit form.

mark9064 commented:
Code size should be OK, the linker strips all unused symbols. A small increase is not an issue :)

Andersama commented:
@mark9064 I'll push, but from what I see of my build, it's oversized. It doesn't appear it'll fit on the pinetime using that appraoch.

Andersama committed:
Use pcg-cpp git repo as submodule

mark9064 commented:
Huh, how has this happened. All other implementations should be stripped...

I'll try take a look, but might be a while honestly

Andersama committed:
Rewrite rng component using <random> as reference

Incorporate interface for <random> engines and std::seed_seq into single struct.
Use operator()() and operator()(uint32_t) to retrieve unbounded and bounded results.

Melissa's PCG is unbiased, so <random> is not required in order to use the RNG struct as is.
Use generate(It start, It end) to fill target range with data (as per std::seed_seq).
Use seed functions with something like std::seed_seq with a generate function to fill the internal state of the PCG with random data.
Or alternatively use the seed function with an integer to select a fairly random PCG stream.

Andersama commented:
@mark9064 I've implemented the interface for `<random>` inside the prng component, so if you want to use the std library's approach you can...although as you saw in the earlier build, there's some strange sizing issues.

DavisNT commented:
@Andersama, @Avamander What do you think about a different approach - reseeding PRNG on every roll of the Dice app (in addition to replacing existing Mersenne Twister with any small footprint PRNG like `std::minstd_rand`)?
See #2478 for details.

Andersama commented:
@DavisNT well, from limited experience, seeding a prng tends to be more expensive than using it. But arguably if you're "properly" seeding the prng, you probably have a decent source of randomness, or you really should. You kind of end up in a catch 22 there, yes ultimately it'd be better than MT, this is one of the reasons I don't like MT at all. It's so expensive on memory, not exactly "fast" (in the context of a dice roller this doesn't matter). Your reseeding has to be a fairly good source of randomness...which you might as well just use, but it'd probably be better than MT (which is the crazy part). That's why I proposed these cheap prngs, once they're seeded, their statistical quality's solid, they're fast, tiny, and evenly distributed. Granted that's over a very long period, which would look pretty random and technically would be "fair", not that anyone would live to see that shake out.

Forgive me for not re-reading the whole thread but from memory, the reason I wouldn't use the ble's cryptographic rng over and over is that its implementation is in fact slow because its using some rng hardware that very slowly accumulates random bits (for its own seeding). Now I don't know if there could be contention for that resource for bluetooth...so I wouldn't hammer on that interface too hard. If I remember correctly the ble implementation memorizes the fact it's been seeded from the cryptographic source and then uses another prng from there. So it could be fine.

It sounds like they're looking at using a different prng that'll be included in a more recent version of GCC, which I think will be a decent choice. I wouldn't have waited this long and it looks like that potential build is larger than this, but hey...I'm not actively wearing this watch...although it is fun to play with.
